backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/incremental/updater/OverlapsUpdater.java

```diff
@@ -0,0 +1,155 @@
+/*******************************************************************************
+ * Copyright (c) 2021 THALES GLOBAL SERVICES.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sirius.web.diagrams.layout.incremental.updater;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import org.eclipse.sirius.web.diagrams.Position;
+import org.eclipse.sirius.web.diagrams.Size;
+import org.eclipse.sirius.web.diagrams.layout.incremental.IncrementalLayoutEngine;
+import org.eclipse.sirius.web.diagrams.layout.incremental.data.IContainerLayoutData;
+import org.eclipse.sirius.web.diagrams.layout.incremental.data.NodeLayoutData;
+import org.eclipse.sirius.web.diagrams.layout.incremental.utils.Bounds;
+import org.eclipse.sirius.web.diagrams.layout.incremental.utils.Geometry;
+
+/**
+ * An algorithm dedicated to solve overlaps issues. Any node in a given container might be moved to avoid overlaps,
+ * except those that have been moved by an user.
+ *
+ * @author wpiers
+ */
+public class OverlapsUpdater {
+
+    private Set<NodeLayoutData> fixedNodes = new HashSet<>();
+
+    public void update(IContainerLayoutData container) {
+        Collection<NodeLayoutData[]> overlaps = this.findAllOverlaps(container);
+        while (!overlaps.isEmpty()) {
```

Comment:
In theory it should not happen since already moved element by the OverlapsUpdater are marked as pinned. Pinned elements are excluded from the next iteration.